evm_arithmetization/src/curve_pairings.rs

```diff
@@ -150,366 +141,937 @@ where
+                        0xbc4b313370b38ef3,
+                        0xec9e99ad690c3395,
+                        0x090689d0585ff075,
+                    ]),
+                },
+            },
+        };
+    }
+
+    /// The sloped line function for doubling a point.
+    pub(crate) fn tangent(p: CurveAff<BN254>, q: CurveAff<Fp2<BN254>>) -> Fp12<BN254> {
+        let cx = -BN254::new(3) * p.x * p.x;
+        let cy = BN254::new(2) * p.y;
+        sparse_embed::<BN254>(p.y * p.y - BN254::new(9), q.x * cx, q.y * cy)
+    }
+
+    /// The sloped line function for adding two points.
+    pub(crate) fn chord(
+        p1: CurveAff<BN254>,
+        p2: CurveAff<BN254>,
+        q: CurveAff<Fp2<BN254>>,
+    ) -> Fp12<BN254> {
+        let cx = p2.y - p1.y;
+        let cy = p1.x - p2.x;
+        sparse_embed::<BN254>(p1.y * p2.x - p2.y * p1.x, q.x * cx, q.y * cy)
+    }
+
+    // The tate pairing takes a point each from the curve and its twist and outputs
+    // an Fp12 element.
+    pub(crate) fn tate(p: CurveAff<BN254>, q: CurveAff<Fp2<BN254>>) -> Fp12<BN254> {
+        let miller_output = miller_loop(p, q);
+        final_exponent(miller_output)
+    }
+
+    /// Standard code for miller loop, can be found on page 99 at this url:
+    /// <https://static1.squarespace.com/static/5fdbb09f31d71c1227082339/t/5ff394720493bd28278889c6/1609798774687/PairingsForBeginners.pdf#page=107>
+    /// where BN_EXP is a hardcoding of the array of Booleans that the loop
+    /// traverses.
+    pub(crate) fn miller_loop(p: CurveAff<BN254>, q: CurveAff<Fp2<BN254>>) -> Fp12<BN254> {
+        let mut r = p;
+        let mut acc: Fp12<BN254> = Fp12::<BN254>::UNIT;
+        let mut line: Fp12<BN254>;
+
+        for i in BN_EXP {
+            line = tangent(r, q);
+            r = r + r;
+            acc = line * acc * acc;
+            if i {
+                line = chord(p, r, q);
+                r = r + p;
+                acc = line * acc;
+            }
+        }
+        acc
+    }
+
+    /// The output y of the miller loop is not an invariant,
+    /// but one gets an invariant by raising y to the power
+    ///     (p^12 - 1)/N = (p^6 - 1)(p^2 + 1)(p^4 - p^2 + 1)/N
+    /// where N is the cyclic group order of the curve.
+    /// To achieve this, we first exponentiate y by p^6 - 1 via
+    ///     y = y_6 / y
+    /// and then exponentiate the result by p^2 + 1 via
+    ///     y = y_2 * y
+    /// We then note that (p^4 - p^2 + 1)/N can be rewritten as
+    ///     (p^4 - p^2 + 1)/N = p^3 + (a2)p^2 + (a1)p - a0
+    /// where 0 < a0, a1, a2 < p. Then the final power is given by
+    ///     y = y_3 * (y^a2)_2 * (y^a1)_1 * (y^-a0).
+    pub(crate) fn final_exponent(f: Fp12<BN254>) -> Fp12<BN254> {
+        let mut y = f.frob(6) / f;
+        y = y.frob(2) * y;
+        let (y_a2, y_a1, y_a0) = get_custom_powers(y);
+        y.frob(3) * y_a2.frob(2) * y_a1.frob(1) * y_a0
+    }
+
+    /// We first together (so as to avoid repeated steps) compute
```

Comment:
```suggestion
    /// In order to avoid repeated steps, we first compute
```
Not sure, but looks better like this